mqtt/src/main/scala/Mqtt.scala

```diff
@@ -9,27 +9,31 @@ import org.eclipse.paho.client.mqttv3.{ MqttMessage => PahoMqttMessage, _ }
 import scala.util._
 import scala.language.implicitConversions
 
-case class MqttSourceSettings(
+final case class MqttSourceSettings(
   connectionSettings: MqttConnectionSettings,
   topics:             Map[String, Int]
 )
 
-case class MqttConnectionSettings(
+final case class MqttConnectionSettings(
   broker:      String,
   clientId:    String,
   persistence: MqttClientPersistence
 )
 
-case class MqttMessage(topic: String, payload: ByteString)
+final case class MqttMessage(topic: String, payload: ByteString)
 
-trait MqttConnectorLogic { this: GraphStageLogic =>
+private[mqtt] trait MqttConnectorLogic { this: GraphStageLogic =>
 
   import MqttConnectorLogic._
 
   def connectionSettings: MqttConnectionSettings
-  def onConnect: AsyncCallback[IMqttAsyncClient]
-  def onMessage: AsyncCallback[MqttMessage]
-  def onConnectionLost: AsyncCallback[Throwable]
+  def handleConnection(client: IMqttAsyncClient): Unit
+  def handleMessage(message: MqttMessage): Unit
+  def handleConnectionLost(ex: Throwable): Unit
+
+  def onConnect = getAsyncCallback[IMqttAsyncClient](handleConnection)
+  val onMessage = getAsyncCallback[MqttMessage](handleMessage)
+  val onConnectionLost = getAsyncCallback[Throwable](handleConnectionLost)
```

Comment:
It's fine to create it like this in the constructor, but you are not allowed to `invoke` it before `preStart`. That is a problem if you expose the callback to the external world (via materialized value).